ext/MooncakeAMDGPUExt.jl

```diff
@@ -0,0 +1,239 @@
+module MooncakeAMDGPUExt
+
+using LinearAlgebra, Random, Mooncake
+
+using Base: IEEEFloat
+using AMDGPU: ROCArray
+
+import Mooncake:
+    MinimalCtx,
+    rrule!!,
+    @is_primitive,
+    tangent_type,
+    fdata_type,
+    rdata_type,
+    primal,
+    tangent,
+    zero_tangent_internal,
+    randn_tangent_internal,
+    increment_internal!!,
+    set_to_zero_internal!!,
+    _add_to_primal_internal,
+    _diff_internal,
+    _dot_internal,
+    _scale_internal,
+    TestUtils,
+    CoDual,
+    NoPullback,
+    to_cr_tangent,
+    increment_and_get_rdata!,
+    MaybeCache,
+    IncCache,
+    NoRData
+
+import Mooncake.TestUtils:
+    populate_address_map_internal, AddressMap, __increment_should_allocate
+
+const ROCFloatArray = ROCArray{<:IEEEFloat}
```

Comment:
Is it possible that we can do a union type to help share code between `CuArray` and `ROCArray`? 

```julia
GPUFloatArray = Union{ROCArray{<:IEEEFloat},CuArray{<:IEEEFloat}}
GPUComplexArray = Union{ROCArray{<:Complex{<:IEEEFloat}},CuArray{<:Complex{<:IEEEFloat}}}
```